cpp/src/arrow/compute/kernels/vector_sort.cc

```diff
@@ -346,14 +374,396 @@ void AddSortingKernels(VectorKernel base, VectorFunction* func) {
+      auto order = sort_key.order;
+      int64_t index_left = 0;
+      auto array_left = sort_key.ResolveChunk<ArrayType>(current_left_, index_left);
+      int64_t index_right = 0;
+      auto array_right = sort_key.ResolveChunk<ArrayType>(current_right_, index_right);
+      if (sort_key.null_count > 0) {
+        auto is_null_left = array_left->IsNull(index_left);
+        auto is_null_right = array_right->IsNull(index_right);
+        if (is_null_left && is_null_right) {
+          return 0;
+        } else if (is_null_left) {
+          return 1;
+        } else if (is_null_right) {
+          return -1;
+        }
+      }
+      auto left = array_left->GetView(index_left);
+      auto right = array_right->GetView(index_right);
+      int32_t compared;
+      if (left == right) {
+        compared = 0;
+      } else if (left > right) {
+        compared = 1;
+      } else {
+        compared = -1;
+      }
+      if (order == SortOrder::DESCENDING) {
+        compared = -compared;
+      }
+      return compared;
+    }
+
+    Status status_;
+    std::vector<ResolvedSortKey> sort_keys_;
+    int64_t current_left_;
+    int64_t current_right_;
+    size_t current_sort_key_index_;
+    int32_t current_compared_;
+  };
+
+ public:
+  TableSorter(uint64_t* indices_begin, uint64_t* indices_end, const Table& table,
+              const SortOptions& options)
+      : indices_begin_(indices_begin),
+        indices_end_(indices_end),
+        comparer_(table, options.sort_keys) {}
+
+  Status Sort() {
+    ARROW_RETURN_NOT_OK(comparer_.status());
+    return comparer_.sort_keys()[0].type->Accept(this);
+  }
+
+#define VISIT(TYPE) \
+  Status Visit(const TYPE##Type& type) override { return SortInternal<TYPE##Type>(); }
+
+  VISIT(Int8)
+  VISIT(Int16)
+  VISIT(Int32)
+  VISIT(Int64)
+  VISIT(UInt8)
+  VISIT(UInt16)
+  VISIT(UInt32)
+  VISIT(UInt64)
+  VISIT(Float)
+  VISIT(Double)
+  VISIT(String)
+  VISIT(Binary)
+  VISIT(LargeString)
+  VISIT(LargeBinary)
+
+#undef VISIT
+
+ private:
+  template <typename Type>
+  Status SortInternal() {
+    using ArrayType = typename TypeTraits<Type>::ArrayType;
+    std::iota(indices_begin_, indices_end_, 0);
+
+    auto& comparer = comparer_;
+    const auto& first_sort_key = comparer.sort_keys()[0];
+    auto nulls_begin = indices_end_;
+    nulls_begin = PartitionNullsInternal<Type>(first_sort_key);
+    std::stable_sort(
+        indices_begin_, nulls_begin,
+        [&first_sort_key, &comparer](uint64_t left, uint64_t right) {
+          int64_t index_left = 0;
+          auto array_left = first_sort_key.ResolveChunk<ArrayType>(left, index_left);
+          int64_t index_right = 0;
+          auto array_right = first_sort_key.ResolveChunk<ArrayType>(right, index_right);
+          auto value_left = array_left->GetView(index_left);
+          auto value_right = array_right->GetView(index_right);
+          if (value_left == value_right) {
+            return comparer.Compare(left, right, 1);
```

Comment:
Performance will probably be very bad if there are lots of equal values in the first column, right?

Reply:
Confirmed. 2x-3x slower in the following cases:

The number of records: 1048576

Wide range: Almost elements will be different

Narrow range: Almost elements may be same

N sort keys   | 2           | 8           | 16
------------- | ----------- | ----------- | -----------
Narrow range  | 373997366ns | 556003297ns | 519211265ns
Wide range    | 167679014ns | 167135310ns | 166726078ns
Narrow / Wide | 2.23        | 3.32        | 3.11

In wide range cases, the number of sort keys don't affect performance.